python/ql/src/Functions/SignatureOverriddenMethod.ql

```diff
@@ -13,23 +13,266 @@
+Function resolveCall(Call call) {
+  exists(DataFlowCall dfc | call = dfc.getNode().(CallNode).getNode() |
+    result = viableCallable(dfc).(DataFlowFunction).getScope()
+  )
+}
+
+/** Holds if `call` may resolve to either `base` or `sub`, and `base` is overridden by `sub`. */
+predicate callViableForEitherOverride(Function base, Function sub, Call call) {
+  overrides(base, sub) and
+  base = resolveCall(call) and
+  sub = resolveCall(call)
+}
+
+/** Holds if either both `base` and `sub` are static methods, or both are not static methods, and `base` is overridden by `sub`. */
+predicate matchingStatic(Function base, Function sub) {
+  overrides(base, sub) and
+  (
+    isStaticmethod(base) and
+    isStaticmethod(sub)
+    or
+    not isStaticmethod(base) and
+    not isStaticmethod(sub)
+  )
+}
+
+int extraSelfArg(Function func) { if isStaticmethod(func) then result = 0 else result = 1 }
+
+/** Holds if the call `call` matches the signature for `func`. */
+predicate callMatchesSignature(Function func, Call call) {
+  func = resolveCall(call) and
+  (
+    // Each parameter of the function is accounted for in the call
+    forall(Parameter param, int i | param = func.getArg(i) |
+      // self arg
+      i = 0 and not isStaticmethod(func)
+      or
+      // positional arg
+      i - extraSelfArg(func) < call.getPositionalArgumentCount()
+      or
+      // has default
+      exists(param.getDefault())
+      or
+      // keyword arg
+      call.getANamedArgumentName() = param.getName()
+    )
+    or
+    // arbitrary varargs or kwargs
+    exists(call.getStarArg())
+    or
+    exists(call.getKwargs())
   ) and
-  not derived.getScope().isSpecialMethod() and
-  derived.getName() != "__init__" and
-  derived.isNormalMethod() and
-  // call to overrides distributed for efficiency
+  // No excess parameters
+  call.getPositionalArgumentCount() + extraSelfArg(func) <= func.getMaxPositionalArguments() and
+  (
+    exists(func.getKwarg())
+    or
+    forall(string name | name = call.getANamedArgumentName() | exists(func.getArgByName(name)))
+  )
+}
+
+/** Gets a call which matches the signature of `base`, but not of overridden `sub`. */
+Call getASignatureMismatchWitness(Function base, Function sub) {
+  callViableForEitherOverride(base, sub, result) and
+  callMatchesSignature(base, result) and
+  not callMatchesSignature(sub, result)
+}
+
+/** Choose a 'witnessing' call that matches the signature of `base` but not of overridden `sub`, and is in the file `file`. */
+Call chooseASignatureMismatchWitnessInFile(Function base, Function sub, File file) {
+  result =
+    min(Call c |
+      c = getASignatureMismatchWitness(base, sub) and
+      c.getLocation().getFile() = file
+    |
+      c order by c.getLocation().getStartLine(), c.getLocation().getStartColumn()
+    )
+}
+
+/** Choose a 'witnessing' call that matches the signature of `base` but not of overridden `sub`. */
+Call chooseASignatureMismatchWitness(Function base, Function sub) {
```

Comment:
There's a pretty bad join in this predicate (which is why the DCA results look kinda terrible). You should be able to fix it by adding this helper predicate and referencing it in all of the appropriate places:
```ql
pragma[nomagic]
private File getFunctionFile(Function f) { result = f.getLocation().getFile() }
```